Added the facility to record local testing accuracies in .csv files

kevinsun203 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
When running a job, if the configuration file has the attribute "do_test" set to true and if there also exists a "results" attribute, then the test accuracies of each client will be computed locally and stored in a csv file with the round number, client ID and test accuracy as headers.
<!--- Describe motivation and context -->
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How has this been tested?
<!--- Please describe in detail how you tested your changes. -->
Tested on local machine using the "fedavg_async_lenet5" configuration file  with 1-3 clients. Each with 1-3 rounds after setting "do_test" to True and another time to False
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue) Fixes #
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

baochunli commented:
Several major issues with this PR:

- The description of the new feature is way too brief, making it difficult to understand what has been added.
- The test accuracies are forced to be logged by all the jobs.
- The log .csv file should always have a fixed number of columns and a variable number of rows, just like the schema design of any database table. The current design is extremely poor.

kevinsun203 commented:
Thank you for the feedback. This is due to my own lack of knowledge but I don't understand what is meant by the test accuracies being forced to be logged by all the jobs. I thought that this was the intention for any configuration file with a "results" attribute.

Since the log.csv is updated after each round of training and since the round numbers have to be the headers of the csv in order to stop the data from being built too wide, I assumed that the data would be logged column by column. I couldn't find a way to log the data row by row with a fixed number of columns.

I'm sorry that this is taking so long, I've never worked with a python repository this big and complex.

cying17 commented:
@kevinsun203 The test accuracies should only be logged when clients compute test accuracy locally.

kevinsun203 commented:
I believed I fixed the issue that forced all jobs to log test accuracies and I've also modified the logging process so that the test accuracy csv file is built row by row. Is this better?

cying17 commented:
This looks better. Thanks. Have you tested when Config().clients.do_test is false?

kevinsun203 commented:
Yes, when it's false, the log accuracy csv is never initialized.

cying17 commented:
Cool, thanks.

baochunli commented on plato/utils/csv_processor.py:
```diff
@@ -23,6 +23,7 @@ def initialize_csv(result_csv_file: str, recorded_items: List,
 def write_csv(result_csv_file: str, new_row: List) -> None:
     """ Write the results of current round. """
-    with open(result_csv_file, 'a', encoding='utf-8') as result_file:
+    with open(result_csv_file, 'a', encoding='utf-8',
```
newline='' needs to be removed.
